### PR TITLE
Add support for user banners/banner color

### PR DESCRIPTION
**Pull request by VelvetThePanda**
Whilst this change has not actually been implemented into the API yet, it was said that there are plans to "start passively serializing banner hashes for bots". Figured I'd open this PR pre-emptively.

### CODE REVIEW

**Comment by Neuheit** [comment]
Does Discord have a PR for this yet?

**Comment by IDoEverything** [comment]
The banner field isn't actually sent currently, this is just pre-emptive for when they do start serializing it.

**Comment by VelvetThePanda** [comment]
It's been confirmed to be in the canary api now (`canary.discord.com/api`), so I'd say this PR should be mergable, since it should trickle down to stable in the near future

**Comment by Lulalaby** [comment]
Here https://github.com/discord/discord-api-docs/pull/3448

**Comment by VelvetThePanda** [comment]
seems to be in the API based on these logs
```fix
[4:26:16 25 AM] [VRB] [FurCord.NET.IRestClient] [Rest ↑] gateway/bot  
[4:26:16 60 AM] [VRB] [FurCord.NET.IRestClient] [Rest ↓] OK {"url": "wss://gateway.discord.gg", "shards": 1, "session_start_limit": {"total": 1000, "remaining": 971, "reset_after": 77731066, "max_concurrency": 1}} 
[4:26:16 80 AM] [VRB] [FurCord.NET.IRestClient] [Rest ↑] users/@me  
[4:26:16 87 AM] [VRB] [FurCord.NET.IRestClient] [Rest ↓] OK {"id": "745325268645707916", "username": "Silk! Canary", "avatar":"2684eecfeb295b0e5dc310e7bae80fa9", "discriminator": "7149", "public_flags": 0, "flags": 0, "bot":true, "banner": null,"banner_color": null, "bio": "", "locale": "en-US", "mfa_enabled": true, "email": null, "verified": true}
```

**Comment by IDoEverything** [comment]
This will not work, properties have to be added to `TransportUser`.

**Comment by VelvetThePanda** [comment]
I never understood, what is the difference between TransportUser and DiscordUser anwyay

**Comment by IDoEverything** [comment]
`TransportUser` is what's actually serialized when receiving a user, and the `DiscordUser` object is constructed from it. This behavior is kinda inconsistent across the library and aims to be cleaned up for 5.x. Also you committed changes to the wrong pr.

**Review by IDoEverything — CHANGES_REQUESTED** [review]
The properties should be overriden in `DiscordMember` and also set in `DiscordUser`'s constructor

File: DSharpPlus/Entities/Guild/DiscordMember.cs [hunk 1]
```diff
@@ -223,6 +223,19 @@ public override string Discriminator
         }
 
         /// <summary>
+        /// Gets the member's banner hash.
+        /// </summary>
+        [JsonIgnore]
+        public override string BannerHash
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Guild/DiscordMember.cs:229]: Also add `BannerUrl`

**Review by IDoEverything — CHANGES_REQUESTED** [review]

File: DSharpPlus/Entities/Guild/DiscordMember.cs (hunk 1 above)
Comment (VelvetThePanda) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:229]: but...why

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:229]: I get it's for consistency but

Comment (IDoEverything) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:229]: You can't access the user object from `DiscordMember` so it should reflect everything

File: DSharpPlus/Net/Abstractions/Transport/TransportUser.cs [hunk 2]
```diff
@@ -39,6 +39,12 @@ internal class TransportUser
         [JsonProperty("avatar", NullValueHandling = NullValueHandling.Ignore)]
         public string AvatarHash { get; internal set; }
 
+        [JsonProperty("banner", NullValueHandling = NullValueHandling.Ignore)]
+        public string BannerHash { get; internal set; }
+
+        [JsonProperty("banner_color")]
```
Comment (Lulalaby) [inline DSharpPlus/Net/Abstractions/Transport/TransportUser.cs:45]:
```suggestion
        [JsonProperty("accent_color")]
```

File: DSharpPlus/Entities/User/DiscordUser.cs [hunk 3]
```diff
@@ -68,12 +69,35 @@ internal DiscordUser(TransportUser transport)
         internal int DiscriminatorInt
             => int.Parse(this.Discriminator, NumberStyles.Integer, CultureInfo.InvariantCulture);
 
+        /// <summary>
+        /// Gets the user's banner color, if set. Mutually exclusive with <see cref="BannerHash"/>.
+        /// </summary>
+        public virtual DiscordColor? BannerColor
+            => this._bannerColor == 0 ? null : new DiscordColor(this._bannerColor);
+
+        [JsonProperty("banner_color")]
```
Comment (Lulalaby) [inline DSharpPlus/Entities/User/DiscordUser.cs:78]:
```suggestion
        [JsonProperty("accent_color")]
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]
Is's accent_color not banner_color

**Comment by VelvetThePanda** [comment]
The API currently sends both, 
 

But Mason says they're currently transitioning

**Comment by Lulalaby** [comment]
It'll be accent_color in future

**Comment by VelvetThePanda** [comment]
> ## But Mason says they're currently transitioning

**Comment by Lulalaby** [comment]
> 
> 
> > ## But Mason says they're currently transitioning

oh ok

**Comment by Lulalaby** [comment]
I had a little problem to fetch banners & co.
For this I implemented into the DiscordClient.cs the var fetch, which forces to refresh a user.
It worked then.
And 
```
                old.BannerHash = usr.BannerHash;
                old._bannerColor = usr._bannerColor;
```
where missing.

```csharp
        public async Task<DiscordUser> GetUserAsync(ulong userId, bool fetch = false)
        {
            if (!fetch && this.TryGetCachedUserInternal(userId, out var usr))
                return usr;

            usr = await this.ApiClient.GetUserAsync(userId).ConfigureAwait(false);
            usr = this.UserCache.AddOrUpdate(userId, usr, (id, old) =>
            {
                old.Username = usr.Username;
                old.Discriminator = usr.Discriminator;
                old.AvatarHash = usr.AvatarHash;
                old.BannerHash = usr.BannerHash;
                old._bannerColor = usr._bannerColor;
                return old;
            });

            return usr;
        }
```

File: DSharpPlus/Entities/User/DiscordUser.cs [hunk 4]
```diff
@@ -68,12 +69,35 @@ internal DiscordUser(TransportUser transport)
...
+        [JsonProperty("banner", NullValueHandling = NullValueHandling.Ignore)]
+        public virtual string BannerHash { get; internal set; }
+
         /// <summary>
         /// Gets the user's avatar hash.
         /// </summary>
         [JsonProperty("avatar", NullValueHandling = NullValueHandling.Ignore)]
         public virtual string AvatarHash { get; internal set; }
 
+
```
Comment (IDoEverything) [inline DSharpPlus/Entities/User/DiscordUser.cs:100]: There's an extra line for some reason

**Comment by IDoEverything** [comment]
The docs pr has been merged, so this should be good to go

File: DSharpPlus/Entities/User/DiscordUser.cs [hunk 5]
```diff
@@ -42,6 +42,7 @@ internal DiscordUser(TransportUser transport)
             this.Username = transport.Username;
             this.Discriminator = transport.Discriminator;
             this.AvatarHash = transport.AvatarHash;
+            this._bannerColor = transport.BannerColor;
```
Comment (IDoEverything) [inline DSharpPlus/Entities/User/DiscordUser.cs:45]: Shouldn't `BannerHash` also be set here

**Review by IDoEverything — APPROVED** [review]

**Pull request merged**